src/createSlice.ts

```diff
@@ -0,0 +1,115 @@
+import { Action, AnyAction, ActionCreator, Reducer } from 'redux'
+import { createAction, PayloadAction } from './createAction'
+import { createReducer, CaseReducersMapObject } from './createReducer'
+import { createSliceSelector, createSelectorName } from './sliceSelector'
+
+export interface Slice<
+  S = any,
+  A extends Action = AnyAction,
+  AT extends string = string
+> {
+  /**
+   * The slice name.
+   */
+  slice: string
+
+  /**
+   * The slice's reducer.
+   */
+  reducer: Reducer<S, A>
+
+  /**
+   * Action creators for the types of actions that are handled by the slice
+   * reducer.
+   */
+  actions: { [type in AT]: ActionCreator<A> }
+
+  /**
+   * Selectors for the slice reducer state. `createSlice()` inserts a single
+   * selector that returns the entire slice state and whose name is
+   * automatically derived from the slice name (e.g., `getCounter` for a slice
+   * named `counter`).
+   */
+  selectors: { [key: string]: (state: any) => S }
```

Comment:
Why is this `any` here? Why does this _return_ `S`? Shouldn't `S` be the input?

This is one object that probably should actually be a generic argument, but annoyingly that kills function argument inference.

Reply:
I fixed this in my version, The `selectors` has a fixed name `getSlice` instead of generating one which messes up auto-complete.

```js
/** 
 * // ...
 * @template SS - [SliceState] - The slice of state the reducer handles
 * @template S - [State] - The redux state tree.
 * // ...
 */
interface Slice<A = any, SS = any, S = any> {
selectors: {
    getSlice: (state: S) => SS;
  };
// ...
}
```
where `S` in

Reply:
> Why is this `any` here? Why does this _return_ `S`? Shouldn't `S` be the input?

The input is the state that _contains_ the state of type `S`. So if the slice name were a type parameter `N` (not sure if it should), the selector type would be `(state: { [key in N]: S }) => S`.